### PR TITLE
Build: Fail if any libs depend on non-core libs

### PR DESCRIPTION
**Pull request by nik9000**
Adds a `precommit` to every subprojects of `:libs` that fails the build
if there are any `:libs` dependencies.

I imagine it'd be possible to fail the build on startup if the
dependencies are unacceptable rather than use a precommit task but the
precommit task reports the error nicely and is easy to skip if for some
reason we want to temporarily skip it. So it felt like the right way to
go.

Since we now have three places where we resolve project substitutions
I've added `dependencyToProject` to `project.ext` in all projects. It
resolves both `project` style dependencies and "external" style (like
"org.elasticsearch:elasticsearch-core:${version}") dependencies to
`Project`s using the `projectSubstitutions`. I use this new function all
three places where resovle project substitutions.

Finally this pulls `apply plugin: 'elasticsearch.build'` out of
`libs/*/build.gradle` and into a subprojects clause in
`libs/build.gradle`. I do this entirely so that I can call
`tasks.precommit.dependsOn checkDependencies` without waiting for the
subprojects to be evaluated or worrying about whether or not they have
`precommit` set up in a normal way.

### CODE REVIEW

**Comment by rjernst** [comment]
I don't think we should do this as a task. This should fail during configuration. We already do similar traversal inside BuildPlugin.configureConfigurations, where we disable transitive dependencies (and always force version conflict failure). I also think this is going to be a little more complicated as we need "groups" of libs that *can* have dependencies on each other (eg painless is going to have painless-compiler and painless-whitelist jars).

**Comment by jasontedor** [comment]
> I also think this is going to be a little more complicated as we need "groups" of libs that can have dependencies on each other (eg painless is going to have painless-compiler and painless-whitelist jars).

I think we can wait to cross that bridge when we get there.

**Comment by rjernst** [comment]
> I think we can wait to cross that bridge when we get there.

I'm in the middle of splitting the painless jars out right now.

**Comment by jasontedor** [comment]
> I'm in the middle of splitting the painless jars out right now.

That is, by my estimation, at least two if not three weeks away. Let us make the "group" change a follow-up to this PR, we can progress when the need is clearer.

**Comment by rjernst** [comment]
That's fine, but my concern about the way this is implemented still stands. Doing it via a task is not right; it needs to be checked during configuration.

**Comment by jasontedor** [comment]
> That's fine, but my concern about the way this is implemented still stands. Doing it via a task is not right; it needs to be checked during configuration.

I have no objections to that, my comment was only about adding the "group" requirement on top of this specific PR.

**Comment by nik9000** [comment]
> I don't think we should do this as a task. This should fail during configuration.

If I think "there are two ways to do this and either are just fine" and I basically flip a coin @rjernst almost always has good arguments against the result of my coin flip. Well, more again the "either are just fine" part of my reasoning, but still. I'll change it to fail on startup.

**Comment by nik9000** [comment]
I've switched this to run during configuration.

**Comment by nik9000** [comment]
@rjernst, would you like to take another look? I've switched this to run during configuration.

File: libs/build.gradle [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+subprojects {
+    /*
+     * All subprojects are java projects using Elasticsearch's standard build
+     * tools.
+     */
+    apply plugin: 'elasticsearch.build'
+
+    /*
+     * Subprojects may depend on the "core" lib but may not depend on any
+     * other Elasticsearch code. This keeps are dependencies simpler to think
```
Comment (rjernst) [inline libs/build.gradle]: `other Elasticsearch code` -> `other lib`

Also, I would reword the second sentence to "This keeps the dependency graph simple"

File: libs/build.gradle [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+     * about.
+     */
+    project.afterEvaluate {
+        configurations.all { Configuration conf ->
+            dependencies.all { Dependency dep ->
+                if (project.name.equals('elasticsearch-nio')) {
+                    System.err.println("ccc $conf.name $dep")
```
Comment (rjernst) [inline libs/build.gradle]: This looks like leftover debugging?

File: libs/build.gradle [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                }
+                Project depProject = dependencyToProject(dep)
+                if (depProject != null
+                        && false == depProject.path.equals(':libs:elasticsearch-core')
+                        && depProject.path.startsWith(':libs')) {
+                    throw new InvalidUserDataException("projects in :libs "
+                            + "may not depend on other projects libs except "
+                            + ":libs:elasticsearch-core but "
+                            + "$project.path depends on $depProject.path")
```
Comment (rjernst) [inline libs/build.gradle]: I think you need brackets around these derefs? `${project.path} depends on ${depProject.path}`

**Review by rjernst — COMMENTED** [review]
This looks much better @nik9000, thanks! I left a few comments.

File: libs/build.gradle (hunk 2 above)
Comment (nik9000) [reply libs/build.gradle]: I'll shoot it.

File: libs/build.gradle (hunk 1 above)
Comment (nik9000) [reply libs/build.gradle]: ++

File: libs/build.gradle (hunk 3 above)
Comment (nik9000) [reply libs/build.gradle]: It isn't required but if you like it better I'm happy to do it that way.

Comment (rjernst) [reply libs/build.gradle]: I think it is clearer that way.
